File: code/lecture-01/java/InvertedIndex.java
// Copyright 2017 University of Freiburg
// Claudius Korzen <korzen@cs.uni-freiburg.de>

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * A simple inverted index as explained in lecture 1.
 */
public class InvertedIndex {
  /**
   * The inverted lists.
   */
  protected Map<String, List<Integer>> invertedLists;

  /**
   * Creates an empty inverted index.
   */
  public InvertedIndex() {
    this.invertedLists = new HashMap<>();
  }

  /**
   * Constructs the inverted index from given file (one record per line).
   * 
   * @param file
   *        The path of the file to process.
   */
  public void readFromFile(String file) {
    try (BufferedReader reader = Files.newBufferedReader(Paths.get(file))) {
      int recordId = 0;
      String line;
      while ((line = reader.readLine()) != null) {
        String[] words = line.split("[^A-Za-z]+");
        for (String word : words) {
          word = word.toLowerCase().trim();

          // Ignore the word if it is empty.
          if (word.isEmpty()) {
            continue;
          }

          // If word seen first time, create inverted list.
          if (!this.invertedLists.containsKey(word)) {
            this.invertedLists.put(word, new ArrayList<>());
          }
          // Append record id to inverted list.
          this.invertedLists.get(word).add(recordId);
        }
        recordId++;
      }
    } catch (IOException e) {
      System.err.println("An error occured on reading the file:");
      e.printStackTrace();
    }
  }

  /**
   * Returns the words within this inverted index.
   */
  public Set<String> getWords() {
    return this.invertedLists.keySet();
  }

  /**
   * Returns the inverted list for the given word. Returns null if this inverted
   * index doesn't contain an inverted list for the word.
   * 
   * @param word
   *        The word to process.
   */
  public List<Integer> getInvertedList(String word) {
    return this.invertedLists.get(word);
  }
}

File: code/lecture-01/python/inverted_index.py
"""
Copyright 2017, University of Freiburg,
Chair of Algorithms and Data Structures.
Author: Hannah Bast <bast@cs.uni-freiburg.de>
"""

import re
import sys


class InvertedIndex:
    """ A very simple inverted index. """

    def __init__(self):
        """ Create an empty inverted index. """

        self.inverted_lists = {}

    def read_from_file(self, file_name):
        """ Construct index from given file.

        >>> ii = InvertedIndex()
        >>> ii.read_from_file("example.txt")
        >>> sorted(ii.inverted_lists.items())
        [('doc', [1, 2, 3]), ('first', [1]), ('second', [2]), ('third', [3])]
        """

        record_id = 0
        with open(file_name) as file:
            for line in file:
                record_id += 1
                for word in re.split("[^a-zA-Z]+", line):
                    if len(word) > 0:
                        word = word.lower()
                        if word not in self.inverted_lists:
                            self.inverted_lists[word] = []
                        # Note that this way, the record ids will be
                        # automaticall in sorted order.
                        self.inverted_lists[word].append(record_id)


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python3 inverted_index.py <file name>")
        sys.exit(1)
    file_name = sys.argv[1]

    ii = InvertedIndex()
    ii.read_from_file(file_name)
    for word, inverted_list in ii.inverted_lists.items():
        print("%s\t%d" % (word, len(inverted_list)))
